### PR TITLE
6 full dataset download pw

### PR DESCRIPTION
**Pull request by philippmwirth**
# Full dataset download

I added the download dataset functionality to the `ApiWorkflowClient` and made it accesible from the CLI. A user can now:

```
# download only the filenames
lightly-download token=123 dataset_id=xyz tag_name=my-tag

# download the full images from the API
lightly-download token=123 dataset_id=xyz tag_name=my-tag output_dir=store/images/here

# download the filenames and then copy the images (more efficient)
lightly-download token=123 dataset_id=xyz tag_name=my-tag output_dir=store/images/here input_dir=path/to/input/images

```

TODO:

- [x] Still need to add tests
- [x] Add a section about this to the documentation

### CODE REVIEW

**Comment by IgorSusmelj** [comment]
Could you please also update the CLI section in the documentation?

File: docs/source/getting_started/command_line_tool.rst [hunk 1]
```diff
@@ -119,16 +119,27 @@ You can upload embeddings directly to the Lightly Platform using the CLI.
...
 -----------------------------------------------
 You can download a dataset with a given tag from the Lightly Platform using the 
-following CLI command. The CLI provides you with two options. Either you 
-download just a list or copy the files from the original dataset into a new 
-folder. The second option is very handy for quick prototyping.
+following CLI command. The CLI provides you with three options:
+
+* Download the list of filenames for a given tag in the dataset.
+  
+* Download the images for a given tag in the dataset.
```
Comment (MalteEbner) [inline docs/source/getting_started/command_line_tool.rst:126]: Here I would add a warning that downloading all images might take long. I prefer saying "Option 2 has the disadvantage that ..." to saying "Option 3 has the advantage of not having the disadvantage of option 2".

File: lightly/api/api_workflow_download_dataset.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    img.save(path)
+
+
+class _DownloadDatasetMixin:
+
+    def download_dataset(self,
+                         output_dir: str,
+                         tag_name: str = 'initial-tag',
+                         verbose: bool = True):
+        """TODO
```
Comment (MalteEbner) [inline lightly/api/api_workflow_download_dataset.py:39]: TODO

File: tests/api_workflow/test_api_workflow_download_dataset.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        with self.assertRaises(ValueError):
+            self.api_workflow_client.download_dataset('path/to/dir', tag_name='this_is_not_a_real_tag_name')
+
+    def test_download_thumbnails(self):
+        def get_thumbnail_dataset_by_id(*args):
+            return DatasetData(name=f'dataset', id='dataset_id', last_modified_at=0,
+                               type='thumbnails', size_in_bytes=-1, n_samples=-1, created_at=-1)
+        self.api_workflow_client.datasets_api.get_dataset_by_id = get_thumbnail_dataset_by_id
+        with self.assertRaises(ValueError):
+            self.api_workflow_client.download_dataset('path/to/dir')
```
Comment (MalteEbner) [inline tests/api_workflow/test_api_workflow_download_dataset.py:41]: The tests are missing a "success" case.

File: lightly/api/api_workflow_download_dataset.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            # only thumbnails or metadata available
+            raise ValueError(
+                f"Dataset with id {self.dataset_id} has no downloadable images!"
+            )
+
+        # check if tag exists
+        available_tags = self._get_all_tags()
+        try:
+            tag = next(tag for tag in available_tags if tag.name == tag_name)
+        except StopIteration:
```
Comment (MalteEbner) [inline lightly/api/api_workflow_download_dataset.py:65]: Why not remove lines 66-68?

File: docs/source/getting_started/command_line_tool.rst (hunk 1 above)
Comment (philippmwirth) [reply docs/source/getting_started/command_line_tool.rst:126]: I like it better this way, has a more positive tone.

**Review by IgorSusmelj — APPROVED** [review]

**Pull request merged**